SRSNV report update

orihir opened the pull request:
Major update to srsnv training report.

Also, allow running on prepared data and trained model (only create report).

orihir committed:
WIP srsnv report

Major update srsnv report.

Also, allow running on prepared data and trained model (only create
report).

orihir committed:
Add shap package for srsnv report

kenissur committed:
"single_read_snv.applicationQC.h5" suffix

orihir committed:
WIP srsnv report fix test bugs

kenissur committed:
ppmSeqStrandVcfAnnotator now handles v7 too

orihir committed:
fix bug with v7 ppmSeq tag annotation

kenissur commented on ugvc/mrd/ppmSeq_utils.py:
```diff
@@ -245,6 +247,27 @@ def process_records(self, records: list[pysam.VariantRecord]) -> list[pysam.Vari
+            if self.adapter_version in (
+                ppmSeqAdapterVersions.V1.value,
+                ppmSeqAdapterVersions.DMBL.value,
+            ):  # legacy_v5_start has start tags only
```
Can you remove this comment please?
# legacy_v5_start has start tags only

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -48,6 +55,123 @@
+
+
+def prob_to_phred(arr, eps=1e-10):
+    return -10 * np.log(1 - arr + eps) / np.log(10)
```
"arr" is the probability of no error, right? As in, arr=[0.9] would return [10], correct?
Please make it clear in the docstring if that's the case

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -48,6 +55,123 @@
+    return 10 * np.log((arr + eps) / (1 - arr + eps)) / np.log(10)
+
+
+def signif(x, p):
```
docstring please, I don't understand this function

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -48,6 +55,123 @@
+    return np.round(x * mags) / mags
+
+
+def list_of_jagged_lists_to_array(list_of_lists, fill=np.nan):
```
docstring

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -48,6 +55,123 @@
+    return discretized_bins
+
+
+# ### Old functions from here
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -141,21 +265,50 @@ def srsnv_report(
 -p output_bepcr_recalls {output_ppmSeq_recalls} \
 -k python3"
-    jupyter_nbconvert_command = f"jupyter nbconvert {reportfile} --output {reporthtml} --to html --no-input"
+    preprocessor_name = "ugvc.mrd.toc_preprocessor.TocPreprocessor"
```
Cool!

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -141,21 +265,50 @@ def srsnv_report(
+
+    # # Write the HTML to a file
+    # with open(reporthtml, 'w') as f:
+    #     f.write(body)
```
remove comments

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -531,9 +699,10 @@ def plot_LoD(
         )
-    plt.xlabel("Base retention ratio on HOM SNVs", fontsize=font_size)
-    plt.ylabel("Measured SNVQ", fontsize=font_size)
+    plt.xlabel("Recall (Base retention ratio on HOM SNVs)", fontsize=font_size)
```
"Recall (Base retention ratio on HOM SNVs)" -> "Recall (retention ratio of TP SNVs)"

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+    lod_filters,
+    lod_label,
+    c_lod,
+    # min_LoD_filter,
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        lod_filters=lod_filters,
+        lod_label=lod_label,
+        c_lod=c_lod,
+        # min_LoD_filter=min_LoD_filter,
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        lod_filters: dict = None,
+        lod_label: dict = None,
+        c_lod: str = "LoD",
+        # min_LoD_filter: str = None,
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+            random_seed = int(datetime.now().timestamp())
+            rng = np.random.default_rng(seed=random_seed)
+            logger.info(f"SRSNVReport: Initializing random number generator with {random_seed=}")
+            self.rng = np.random.default_rng(seed=14)
```
This line seems wrong. I think you should replace these lines:
```
            self.rng = np.random.default_rng(seed=14)
        else:
            self.rng = rng
```

With:
`self.rng = rng`

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        else:
+            self.rng = rng
+
+        if statistics_json_file:
```
But not the other way around?

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        self.data_df["ML_logit_train"] = prob_to_logit(self.data_df["ML_prob_1_train"])
+
+    def _save_plt(self, output_filename: str = None, fig=None, tight_layout=True, **kwargs):
+        if output_filename is not None:
```
I suggest making it a non keyword argument and removing this if clause, this function does nothing if it is not provided

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        # Info about versions
+        version_info = {
+            ("Pipeline version", ""): self.params["pipeline_version"],
+            ("Docker image", ""): self.params["docker_image"],
```
handle the cropping issue

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+
+    @exception_handler
+    def calc_training_metrics_table(self):
+        """Calculte main run quality statistics"""
```
remove if unused

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        # col_order = [
+        #     'qual_FP', 'qual_TP', 'qual_TP_mixed', 'qual_TP_non-mixed',
+        #     'ML_qual_FP', 'ML_qual_TP', 'ML_qual_TP_mixed', 'ML_qual_TP_non-mixed'
+        # ]
```
remove comments

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        if qual_stat_ps is None:
+            qual_stat_ps = [0.05, 0.1, 0.25, 0.5, 0.75, 0.90, 0.95]
+        if cols_for_stats is None:
+            cols_for_stats = {"qual": "qual", "ML_qual_1_test": "ML_qual", "ML_logit_test": "ML_logit"}
```
I suggest defining "qual" and "ML_qual" as constants if they are not defined already, and using that instead of strings

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        """Generate tables of median quality and data quantity per start and end ppmseq tags."""
+        data_df_tp = self.data_df[self.data_df["label"]].copy()
+        ppmseq_tags_in_data = self.start_tag_col is not None and self.end_tag_col is not None
+        start_tag_col, end_tag_col = (self.start_tag_col, self.end_tag_col) if ppmseq_tags_in_data else ("st", "et")
```
Use constants for "st" and "et", defined in ppmSeq_utils

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        ppmseq_category_quality_table = (
+            data_df_tp.groupby([start_tag_col, end_tag_col], dropna=False)["qual"].median().unstack()
+        )
+        if "END_UNREACHED" in ppmseq_category_quality_table.index:
```
"END_UNREACHED" also has a constant in ppmSeq_utils

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+            data_df_tp.groupby([start_tag_col, end_tag_col], dropna=False)["qual"].median().unstack()
+        )
+        if "END_UNREACHED" in ppmseq_category_quality_table.index:
+            ppmseq_category_quality_table = ppmseq_category_quality_table.drop(index="END_UNREACHED")
```
"END_UNREACHED" also has a constant in ppmSeq_utils

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+            data_df_tp.groupby([start_tag_col, end_tag_col], dropna=False)["qual"].count().unstack()
+        )
+        if "END_UNREACHED" in ppmseq_category_quantity_table.index:
+            ppmseq_category_quantity_table = ppmseq_category_quantity_table.drop(index="END_UNREACHED")
```
"END_UNREACHED" also has a constant in ppmSeq_utils

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        ax = axes[0]
+        train_logloss = list_of_jagged_lists_to_array(
+            [result["validation_0"]["mlogloss"] for result in training_results]
+        )  # np.array([result['validation_0']['mlogloss'] for result in training_results])
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        val_logloss = list_of_jagged_lists_to_array(
+            [result["validation_1"]["mlogloss"] for result in training_results]
+        )  # np.array([result['validation_1']['mlogloss'] for result in training_results])
+        # print(train_logloss.shape, val_logloss.shape)
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+    def _get_trinuc_stats(self, q1: float = 0.1, q2: float = 0.9):
+        data_df = self.data_df.copy()
+        data_df["is_cycle_skip"] = data_df["is_cycle_skip"].astype(int)
+        trinuc_stats = data_df.groupby(["trinuc_context_with_alt", "label", "is_forward", "is_mixed"]).agg(
```
These should all be constants too:
"is_cycle_skip", "trinuc_context_with_alt", "label", "is_forward", "is_mixed"

kenissur replied:
Many or all of them already are, just need to import them

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        if order == "symmetric":
+            trinuc_ref_alt = [
+                c1 + r + c2 + a
+                for r, a in (("A", "C"), ("A", "G"), ("A", "T"), ("C", "G"), ("C", "T"), ("G", "T"))
```
Shouldn't the last one be "C", "A"? In fact, I think this is the line you wanted:
`                for r, a in (("A", "C"), ("A", "G"), ("A", "T"), ("C", "A"), ("C", "G"), ("C", "T"))`

kenissur replied:
I might not have understood something but it looks like an anomaly

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+                if r != a
+            ] + [
+                c1 + r + c2 + a
+                for a, r in (("A", "C"), ("A", "G"), ("A", "T"), ("C", "G"), ("C", "T"), ("G", "T"))
```
Again with the "C", "A" as above

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        elif order == "reverse":
+            trinuc_ref_alt = [
+                c1 + r + c2 + a
+                for r, a in (("A", "C"), ("A", "G"), ("A", "T"), ("C", "G"), ("C", "T"), ("G", "T"))
```
`                for r, a in (("A", "C"), ("A", "G"), ("A", "T"), ("G", "A"), ("G", "C"), ("G", "T"))`

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+                if r != a
+            ] + [
+                c1 + r + c2 + a
+                for r, a in (("T", "G"), ("T", "C"), ("T", "A"), ("G", "C"), ("G", "A"), ("C", "A"))
```
`                for r, a in (("A", "C"), ("A", "G"), ("A", "T"), ("G", "A"), ("G", "C"), ("G", "T"))`

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+    def plot_quality_histogram(self, plot_interpolating_function: bool = False, output_filename: str = None):
+        """Plot a histogram of qual values for TP reads, both mixed and non-mixed."""
+        # label_fontsize = 12
+        # ticklabelsfontsize = 12
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
-        max_score,
-        output_filename=output_ppmSeq_recalls,
-    )
+    # ### HERE
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        # LoD_params["specificity_at_lod"] = 0.99
+        # LoD_params["simulated_signature_size"] = 10_000
+        # LoD_params["simulated_coverage"] = 30
+        # LoD_params["minimum_number_of_read_for_detection"] = 2
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+            #     simulated_coverage=LoD_params["simulated_coverage"],
+            #     minimum_number_of_read_for_detection=LoD_params["minimum_number_of_read_for_detection"],
+            #     output_dataframe_file=self.mrd_simulation_dataframe_file,
+            # )
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        #     cls_features,
+        #     df_X_with_pred_columns,
+        #     output_filename=output_qual_per_feature,
+        # )
```
remove comment

kenissur commented on ugvc/mrd/srsnv_plotting_utils.py:
```diff
@@ -1275,210 +1457,1342 @@ def calculate_lod_stats(
+        #     output_filename=output_qual_per_feature,
+        # )
+
+        is_mixed_flag = "is_mixed" in df_X_with_pred_columns
```
constant

kenissur commented on ugvc/mrd/srsnv_training_utils.py:
```diff
@@ -899,21 +931,156 @@ def save_model_and_data(self):
+        cram_tags_in_categorical = "st" in self.categorical_features_names and "et" in self.categorical_features_names
+        if cram_tags_in_categorical and strand_ratio_in_categorical:
+            logger.warning("categorical_features_dict contains both legacy_v5 and v1 ppmSeq tags")
+        if self.ppmSeq_adapter_version == "v1":
```
adapter versions are also defined as constants in ppmSeq_utils

kenissur commented on ugvc/mrd/srsnv_training_utils.py:
```diff
@@ -899,21 +931,156 @@ def save_model_and_data(self):
+    def add_is_mixed_to_featuremap_df(self):
+        """Add is_mixed column to self.featuremap_df"""
+        logger.info("Adding is_mixed column to featuremap")
+        # TODO: use the information from adapter_version instead of this patch
```
is it still a TODO?

kenissur commented on ugvc/mrd/srsnv_training_utils.py:
```diff
@@ -899,21 +931,156 @@ def save_model_and_data(self):
+            #     df_mrd_simulation=self.df_mrd_simulation,
+            #     output_h5=self.test_statistics_h5_file,
+            #     lod_column=self.c_lod,
+            # )
```
remove comment

kenissur commented on ugvc/mrd/srsnv_training_utils.py:
```diff
@@ -935,39 +1102,49 @@ def add_predictions_to_featuremap_df(self, return_train=True):
+        #     featuremap_df[dataset].drop(columns=[f"{col}_{other_dataset}" for col in pred_cols], inplace=True)
+        #     featuremap_df[dataset].rename(columns={f"{col}_{dataset}": col for col in pred_cols}, inplace=True)
+        #     # Make sure to take only reads that are indeed in the "train"/"test" sets:
+        #     featuremap_df[dataset] = featuremap_df[dataset].loc[~featuremap_df[dataset]["ML_prob_1"].isna(), :]
```
remove comment

kenissur commented on ugvc/mrd/srsnv_training_utils.py:
```diff
@@ -935,39 +1102,49 @@ def add_predictions_to_featuremap_df(self, return_train=True):
+            # ],
+            # [self.test_statistics_h5_file, self.train_statistics_h5_file],
+            # [self.test_statistics_json_file, self.train_statistics_json_file],
+            # ["test", "train"],
```
remove comment

kenissur commented on ugvc/mrd/srsnv_training_utils.py:
```diff
@@ -935,39 +1102,49 @@ def add_predictions_to_featuremap_df(self, return_train=True):
+        #     # Make sure to take only reads that are indeed in the "train"/"test" sets:
+        #     featuremap_df[dataset] = featuremap_df[dataset].loc[~featuremap_df[dataset]["ML_prob_1"].isna(), :]
+
+        for (df, statistics_h5_file, statistics_json_file) in zip(
```
Looks like you don't need that loop

kenissur commented on ugvc/mrd/srsnv_training_utils.py:
```diff
@@ -935,39 +1102,49 @@ def add_predictions_to_featuremap_df(self, return_train=True):
+                lod_filters=self.lod_filters_filtered,
+                lod_label=self.lod_label,
+                c_lod=self.c_lod,
+                # min_LoD_filter=self.min_LoD_filter,
```
remove comment

kenissur commented on ugvc/mrd/srsnv_training_utils.py:
```diff
@@ -1070,40 +1247,68 @@ def create_dataframes(self):
+                "train_statistics_h5_file",
+                "test_statistics_json_file",
+                "train_statistics_json_file",
+                "lod_filters",
```
should probably be constants too

kenissur commented on ugvc/pipelines/mrd/srsnv_training.py:
```diff
@@ -183,6 +213,21 @@ def parse_args(argv: list[str]) -> argparse.Namespace:
+        help="""If flag is provided, skip dataset genenration and model training.
+        Output folder should contain featuremap_df.parquet and model.joblib,
+        and all other relevant files.""",
+    )
```
Wouldn't it be better to give the model and dataset paths as input rather than require that they already be saved in the right path?

kenissur commented on ugvc/pipelines/mrd/srsnv_training.py:
```diff
@@ -300,11 +359,11 @@ def run(argv: list[str]):
+    #     model_file=s.model_joblib_save_path,
+    #     params_file=s.params_save_path,
+    #     simple_pipeline=None,
+    # )
```
remove comment

kenissur commented on ugvc/vcfbed/variant_annotation.py:
```diff
@@ -156,18 +156,36 @@ def process_vcf(
+                    variant_count = 1
+                    # Iterate through variants in the contig and count them
+                    # for _ in input_variant_file.fetch(contig):
+                    #     variant_count += 1
```
remove comment

kenissur commented on ugvc/vcfbed/variant_annotation.py:
```diff
@@ -156,18 +156,36 @@ def process_vcf(
+                    # for _ in input_variant_file.fetch(contig):
+                    #     variant_count += 1
+
+                    if variant_count > 0:
```
I don't understand these two lines, seems to me like they do nothing:
```
variant_count = 1
if variant_count > 0:
```

What was wrong with the previous version?

kenissur commented on ugvc/vcfbed/variant_annotation.py:
```diff
@@ -156,18 +156,36 @@ def process_vcf(
+                    else:
+                        logger.info(f"Skipping empty contig {contig}. No variants found.")
+                except StopIteration:  # as e:
+                    skipped_contigs.append(contig)
```
nice

kenissur commented on ugvc/vcfbed/variant_annotation.py:
```diff
@@ -156,18 +156,36 @@ def process_vcf(
+                    # logger.info(f"Skipping empty contig {contig}. Error: {e}")
+            # for contig, count in variant_counts.items():
+            #     if count>0:
+            #         logger.info(f"Contig {contig} has {count} variants.")
```
remove comment

kenissur commented on ugvc/reports/srsnv_report.ipynb:
```diff
@@ -218,37 +474,230 @@
+   "metadata": {},
+   "source": [
+    "---\n",
+    "# Remnants from old report"
```
remember to remove "Remnants from old report"

kenissur reviewed (approved):
Hi Ori, great work!

Approving so I don't delay you, I had a few substance comments and a ton of style ones, please finish what you can of these, and we will fix the others in the next iteration.

One major point about the output h5 file - it needs to contain a Series called keys_to_convert that will tell Papyrus which tables it should read. That would be any simple table, but not complex data like the Shap scores for individual points or the histogram per trinuc (I think). Could you please add that?
After you do we should send an example h5 to Zohar Gan so he can validate it is read to Papyrus properly.

General style comments:
1. Please use string constants. Most of the FeatureMap/ppmSeq stuff are already defined in enumerations, you can import them, you might need to define a few things.
2. Please add a docstring with the regular structure to all of you functions, including typehints
3. Please remove all the commented code

orihir committed:
SRSNV training reserve one chrom for test set

When using cross validation, chromosomes are partitioned into training folds.
Now, a configurable number of chromosomes (by default 1)
is excluded from all folds, and reserved exclusively for the test set.

orihir committed:
SRSNV report add training metrics

Add ML_qual/logit histograms, roc auc scores, variance among k-fold CV models.

Cleanup: remove unused plots from report.

orihir committed:
SRSNV report reorder and add training metrics

And also is_mixed + ML_qual in LoD_plot

orihir committed:
SRSNV report fix auc bug and code cleanup

ROC AUC can only be calculated when the dataset is not empty,
and when both classes (True, False) are represented in the dataset.
This commit adds checks for these conditions before calculation
or ROC AUC. It also adds check whether any reads are mixed before
calculating mixed-only statistics.

In addition, the code for SRSNV report is cleaned from unnecessary
code belonging to old report.

orihir committed:
SRSNV report adapt unit tests, small bug fixes

orihir committed:
SRSNV report update plotting utils

orihir committed:
SRSNV update applicationQC.h5 file

orihir committed:
SRSNV report roc auc bug fix

When calculating roc auc on holdout set, model predictions
may be empty (e.g, for is_mixed set when there are no mixed reads).
In this case, an error was raised when calling predict_proba(...)[:,1] (for
the probabilities of the 1 class). This is now fixed.
